processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java

```diff
@@ -0,0 +1,90 @@
+
+import java.util.UUID;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import com.redhat.service.smartevents.infra.auth.AbstractOidcClient;
+import com.redhat.service.smartevents.infra.exceptions.definitions.user.CloudEventDeserializationException;
+import com.redhat.service.smartevents.infra.exceptions.definitions.user.CloudEventSerializationException;
+import com.redhat.service.smartevents.processor.actions.ActionInvoker;
+
+import io.vertx.core.json.JsonObject;
+import io.vertx.mutiny.core.buffer.Buffer;
+import io.vertx.mutiny.ext.web.client.HttpRequest;
+import io.vertx.mutiny.ext.web.client.WebClient;
+
+public class SourceActionInvoker implements ActionInvoker {
+
+    public static final String CLOUD_EVENT_SOURCE = "RHOAS";
+
+    private static final Logger LOG = LoggerFactory.getLogger(SourceActionInvoker.class);
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+    private static final Supplier<String> UUID_GENERATOR = () -> UUID.randomUUID().toString();
+
+    private final String endpoint;
+    private final String cloudEventType;
+    private final WebClient client;
+    private final AbstractOidcClient oidcClient;
+    private final Supplier<String> idGenerator;
+
+    public SourceActionInvoker(String endpoint, String cloudEventType, WebClient client, AbstractOidcClient oidcClient) {
+        this(endpoint, cloudEventType, client, oidcClient, UUID_GENERATOR);
+    }
+
+    SourceActionInvoker(String endpoint, String cloudEventType, WebClient client, AbstractOidcClient oidcClient, Supplier<String> idGenerator) {
+        this.endpoint = endpoint;
+        this.cloudEventType = cloudEventType;
+        this.client = client;
+        this.oidcClient = oidcClient;
+        this.idGenerator = idGenerator;
+    }
+
+    @Override
+    public void onEvent(String event) {
+        LOG.info("SourceActionInvoker::onEvent | {} - {} | {}", endpoint, cloudEventType, event);
+
+        HttpRequest<Buffer> request = client.postAbs(endpoint);
+        String token = oidcClient.getToken();
+        LOG.info("SourceActionInvoker::onEvent | Token: {}", token);
+        if (token != null && !"".equals(token)) {
+            LOG.info("SourceActionInvoker::onEvent | Token configured");
+            request = request.bearerTokenAuthentication(token);
+        }
+        request.sendJsonObject(getPayload(event)).subscribe().with(
+                response -> LOG.info("RESPONSE {} {}", response.statusCode(), response.statusMessage()),
+                error -> LOG.error("ERROR", error));
+    }
+
+    private JsonObject getPayload(String event) {
+        try {
+            ObjectNode cloudEvent = MAPPER.createObjectNode();
```

Comment:
@kostola Do you have an example of the Event put onto the Kafka Topic by the Slack Source Connector?

I'm interested to understand if the Event contains headers, because at the minute we're putting the Kafka Event headers into the `data` part of the `CloudEvent`. So from a user experience perspective, if I want to match on a header from a Source, I'm interested to understand if they have to write a filter that targets the `data` element of the `CloudEvent`.

e.g. if the Kafka Event produced looks like this:

```
{
 "header_1":"foo"
}
```
Do I as a user have to write the following filter to match on it:

```
{
  "filters": [
    {
      "type": "StringEquals", 
      "key": "data.header_1",
      "value": "foo"
    }
  ]
}
```

Does that match user expectations or would you as a user expect the headers to be hoisted as custom attributes on the `CloudEvent`? 🤔

Reply:
With the refactoring [described here](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/664#discussion_r862846024) now the event received from a source Managed Connector is wrapped in a `CloudEvent` before reaching the action invoker, so the filtering is consistent betweek source and sink processors.

I honestly have no idea about Kafka Event headers. We are totally not dealing with them now.

Imho this should be discussed separately.

Reply:
@kostola @lucamolteni I'm pretty sure this will impact on the S3 Source Connector (and maybe other connectors). From what I can see of the payload shared by @lucamolteni last week, the S3 Connector is using Kafka headers to store some of the data about the event emitted from S3. So we need to understand how we're going to bind the Kafka headers used by the Connector into the wrapped `CloudEvent` so that a user can filter on them in their `Processor`. For example, the Events produced by the S3 Connector seem to have the following headers:

```
Headers: {CamelAwsS3BucketName=test-connector-1, CamelAwsS3ContentControl=null, CamelAwsS3ContentDisposition=null, CamelAwsS3ContentEncoding=null, CamelAwsS3ContentLength=59650, CamelAwsS3ContentType=image/jpeg, CamelAwsS3ETag="4721d67d808c4854581fe119a914564e", CamelAwsS3ExpirationTime=null, CamelAwsS3Key=side-monkey.jpeg, CamelAwsS3LastModified=2022-04-29T13:16:01Z, CamelAwsS3Metadata={}, CamelAwsS3ReplicationStatus=null, CamelAwsS3ServerSideEncryption=null, CamelAwsS3StorageClass=null, CamelAwsS3VersionId=null, CamelMessageTimestamp=1651238161000}
```

If I as user wanted to write a `Filter` targeting the `CamelAwsS3Key`, how would I do that? 

My instinct is that we make these headers Extension Context Attributes (https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#extension-context-attributes) and then place the event payload into the `data` element of the `CloudEvent` as we do now. So taking the S3 Connector payload as an example, we'd end up with a `CloudEvent` looking like:

```
{
  "id" : "blah",   
  "source" : "RHOSE",     
  "specversion" : "1.0",
  "camelawss3bucketname": "test-connector-1",   <-- headers
  "camelawss3contentcontrol": "null", <-- headers
  ...
  "data": {
  }
}
```
Then as a user, I can write my `Filter` as follows:

```
{
  "filters": [
    {
      "type": "StringEquals", 
      "key": "camelawss3key",
      "value": "foo"
    }
  ]
}
```

@lucamolteni Can you confirm I'm correct on the structure of the Kafka Event from S3 Connector please?

So basically rule would be:

- If Source Connector uses Kafka Headers, these are hoisted to Extension Context Attributes
- Payload from the Source Connector is preserved as-is into the `data` part of the `CloudEvent`

cheers,

Rob

cc @danielezonca

Reply:
@danielezonca @lucamolteni @robpblake I think this discussion is getting big enough to require a dedicated task...wdyt?

Reply:
Thanks @kostola @lucamolteni 👍